### DISCUSSION TITLE
Array items: same type

### DISCUSSION THREAD

**Discussion by mfulton26** [json-schema-org/community#685] [Q&A] [answered]
How can a schema be defined where an array's items can be of any type but they must all be of the same type?

**Comment by gregsdennis** [comment]
You're thinking of an open generic array type.  In C#, this would be something like `List<T>`.

I wrote a [blog post](https://json-schema.org/blog/posts/dynamicref-and-generics) on a way to support generic types in JSON Schema a few months ago.  It involves `$dynamicAnchor` and `$dynamicRef`, so you'll need 2020-12.

There's not a way to leave it open (not define the `T`) and validate that all of the items are the same, unfortunately.  You have to define what the item looks like at some point.

**Comment by gregsdennis** [comment, marked as answer]
Another option, if you have a limited set of things an item _could_ be is to use an `anyOf` with a nested `items` subschema for each kind of thing you have.

```jsonc
{
  "type": "array",
  "anyOf": [
    { "items": { "$ref": "#/$defs/item1" } },
    { "items": { "$ref": "#/$defs/item2" } },
    { "items": { "$ref": "#/$defs/item3" } },
    // ...
  ],
  "$defs": {
    "item1": { /* ... */ },
    "item2": { /* ... */ },
    "item3": { /* ... */ },
    // ...
  }
}
```

This would ensure that all items are the same type, but it could be any of the listed types.
